requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,121 @@
+
+## Solution
+
+Scheduling feature flag changes allows users to automatically change flag properties at a future point in time. For example, a user can schedule an initial release of the flag for internal testing, then enable it to the beta users group 2 days later, and finally roll it out to 100% of the users in 2 weeks.
+
+## Scope
+
+This feature should allow the following scheduled modifications to a feature flag:
+
+-   **Enable/disable the flag** (field: `is_active`)
+-   **Add rollout conditions to the flag** (field: `filters`)
+
+We want to be able to _generalize_ this implementation to other models such as `Survey` or `Experiment`.
+
+## Implementation
+
+### API
+
+To schedule future changes, we can have a scheduling API, specifying the model to operate on, the id, the changes to be executed, and the schedule time.
+
+```
+POST /api/projects/:project_id/schedule
+{
+    "model": "FeatureFlag",
+    "id": 57,
+    "changes": [ ... ]
+    "scheduled_at": <timestamp>
+}
+```
+
+A scheduled change can be expressed as a JSON, specifying the operation to be performed and any payload required for the operation. In the context of scheduled changes to feature flags, the operation will be a `db_update`. For other scheduling use cases in the future, the operation may be again a database update or some other side effect.
+
+**Example 1: Enable the flag**
+
+```
+{
+    "operation": "db_update",
+    "payload": {
+        "field": "is_active",
+        "value": "false"
+    }
+}
+```
+
+**Example 2: Add a rollout condition to the flag**
+
+```
+{
+    "operation": "db_update",
+    "payload": {
+        "field": “filter”,
+        "value": {
+            "variant": null,
+            "properties": [{
+                "key": "$geoip_country_name",
+                "type": "person",
+                "value": ["Australia"],
+                "operator": "exact"
+            }],
+            "rollout_percentage": 20
+        }
+    }
+}
+```
+
+### Queue processing
+
+_At_ or _after_ the time specified, the change needs to be propagated onto the record.
+
+We already use Celery with a Redis queue to execute various background tasks. However, because the Redis queue is not durable, we cannot schedule tasks at the precise point in the future. A possible solution here is to have Postgres table of scheduled changes. Celery could periodically poll this table, e.g. every minute. During each poll, Celery would process any rows where the _scheduled_at_ time has already passed.
+
+The `scheduled_changes` table would have the following schema:
+
+| team_id | record_id | model       | changes (JSON) | scheduled_at (timestamp) | executed_at (timestamp, null) |
```

Comment:
implementation reminder that record_id needs to be string to support both int & string model ids

Reply:
below you mention recording failure, that needs to go in the model too!